jumanji/environments/packing/job_shop/generator_test.py

```diff
@@ -103,3 +105,101 @@ def test_random_generator__call(self, random_generator: RandomGenerator) -> None
 
         state2 = call_fn(key=jax.random.PRNGKey(2))
         assert_trees_are_different(state1, state2)
+
+
+class TestMakespanGenerator:
+    NUM_JOBS = 5
+    NUM_MACHINES = 3
+    MAKESPAN = 12
+    MAX_NUM_OPS = MAKESPAN
+    MAX_OP_DURATION = MAKESPAN
+
+    @pytest.fixture
+    def dense_schedule_generator(self) -> DenseGenerator:
+        return DenseGenerator(
+            num_jobs=self.NUM_JOBS,
+            num_machines=self.NUM_MACHINES,
+            max_num_ops=self.MAX_NUM_OPS,
+            max_op_duration=self.MAX_OP_DURATION,
+            makespan=self.MAKESPAN,
+        )
+
+    @pytest.fixture
+    def hardcoded_schedule(self) -> chex.Array:
+        return jnp.array(
+            [
+                [1, 0, 1, 1, 2, 3, 4, 0, 1, 3, 2, 3],
+                [4, 1, 2, 0, 3, 4, 2, 3, 2, 2, 3, 2],
+                [0, 2, 3, 4, 0, 0, 0, 2, 3, 1, 0, 0],
+            ]
+        )
+
+    def test_dense_schedule_generator__attributes(
+        self, dense_schedule_generator: DenseGenerator
+    ) -> None:
+        """Validate that the random instance generator has the correct properties."""
+        assert dense_schedule_generator.num_jobs == self.NUM_JOBS
+        assert dense_schedule_generator.num_machines == self.NUM_MACHINES
+        assert dense_schedule_generator.max_num_ops == self.MAKESPAN
+        assert dense_schedule_generator.max_op_duration == self.MAKESPAN
+        assert dense_schedule_generator.makespan == self.MAKESPAN
+
+        key = jax.random.PRNGKey(0)
+        state = dense_schedule_generator(key)
+        assert isinstance(state, State)
+
+    def test_dense_schedule_generator__generate_schedule(
+        self, dense_schedule_generator: DenseGenerator
+    ) -> None:
+        key = jax.random.PRNGKey(0)
+        schedule = dense_schedule_generator._generate_schedule(key)
+        assert schedule.shape == (self.NUM_MACHINES, self.MAKESPAN)
+        assert jnp.all((schedule >= 0) & (schedule < self.NUM_JOBS))
```

Comment:
Maybe we could add a check that no column contains twice the same value ? something like
```python
        _, counts = jnp.unique(schedule, return_counts=True, axis=0)
        assert jnp.all(counts <= 1)
```

Reply:
Good idea, have added this check to the test 👍